ksql-engine/src/main/java/io/confluent/ksql/planner/plan/FlatMapNode.java

```diff
@@ -0,0 +1,139 @@
+
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.analyzer.TableFunctionAnalysis;
+import io.confluent.ksql.execution.builder.KsqlQueryBuilder;
+import io.confluent.ksql.execution.context.QueryContext;
+import io.confluent.ksql.execution.function.UdtfUtil;
+import io.confluent.ksql.execution.plan.SelectExpression;
+import io.confluent.ksql.function.FunctionRegistry;
+import io.confluent.ksql.function.KsqlTableFunction;
+import io.confluent.ksql.metastore.model.KeyField;
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.schema.ksql.Column;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SchemaConverters.ConnectToSqlTypeConverter;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.services.KafkaTopicClient;
+import io.confluent.ksql.structured.SchemaKStream;
+import java.util.List;
+import java.util.Objects;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * A node in the logical plan which represents a flat map operation - transforming a single row
+ * into zero or more rows.
+ */
+@Immutable
+public class FlatMapNode extends PlanNode {
+
+  private final PlanNode source;
+  private final LogicalSchema schema;
+  private final TableFunctionAnalysis tableFunctionAnalysis;
+
+  public FlatMapNode(
+      final PlanNodeId id,
+      final PlanNode source,
+      final LogicalSchema schema,
+      final TableFunctionAnalysis tableFunctionAnalysis
+  ) {
+    super(id, source.getNodeOutputType());
+    this.source = Objects.requireNonNull(source, "source");
+    this.schema = Objects.requireNonNull(schema);
+    this.tableFunctionAnalysis = Objects.requireNonNull(tableFunctionAnalysis);
+  }
+
+  @Override
+  public LogicalSchema getSchema() {
+    return schema;
+  }
+
+  @Override
+  public KeyField getKeyField() {
+    return source.getKeyField();
+  }
+
+  @Override
+  public List<PlanNode> getSources() {
+    return ImmutableList.of(source);
+  }
+
+  public PlanNode getSource() {
+    return source;
+  }
+
+  @Override
+  public List<SelectExpression> getSelectExpressions() {
+    return tableFunctionAnalysis.getFinalSelectExpressions();
+  }
+
+  @Override
+  public <C, R> R accept(final PlanVisitor<C, R> visitor, final C context) {
+    return visitor.visitFlatMap(this, context);
+  }
+
+  @Override
+  protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
+    return source.getPartitions(kafkaTopicClient);
+  }
+
+  @Override
+  public SchemaKStream<?> buildStream(final KsqlQueryBuilder builder) {
+
+    final QueryContext.Stacker contextStacker = builder.buildNodeContext(getId().toString());
+
+    final LogicalSchema outputSchema = buildLogicalSchema(
+        getSchema(),
+        builder.getFunctionRegistry(),
+        tableFunctionAnalysis
+    );
+
+    return getSource().buildStream(builder).flatMap(outputSchema,
```

Comment:
(this comment doesn't really belong _here_, but I don't know where else to put it). I think pairing the flatMap in the middle of the selects may also cause us some performance grief. Imagine I had:
```sql
SELECT myUdtf(foo), myReallyExpensiveUdf(bar) FROM myStream;
```
Instead of applying `myReallyExpensiveUdf(bar)` once, we're applying it on every single output of `myUdtf(foo)`. If we tie the flatMap as part of the selects, I think it'll be a lot more difficult to reason about pushing the execution of `myReallyExpensiveUdf` up the stack.

Reply:
Talked offline about this one - I think we can handle by changing the way that we arrange the flatMap/project nodes. This can be done irrespective of the approach we choose, and can probably be left to a future PR.